### DISCUSSION TITLE
Including stylesheet for MathML fonts at build time instead of inline in the docs source

### DISCUSSION THREAD

**Discussion by schalkneethling** [mdn/mdn-community#260] [Platform] [answered]
Currently on MDN Web Docs we have a polyfill that is loaded by Yari if it determines that MathML is not supported in the current browser. Here is the code that does this.

- https://github.com/mdn/yari/tree/80a3a9876b41d903d678636d88cb6d84a100c7b2/client/src/document/mathml-polyfill
- The CSS: https://github.com/mdn/yari/blob/80a3a9876b41d903d678636d88cb6d84a100c7b2/client/src/document/mathml-polyfill/mathml.css
- The font: https://github.com/mdn/yari/tree/80a3a9876b41d903d678636d88cb6d84a100c7b2/client/src/document/mathml-polyfill/font

Now, on a number of MathML pages we also have another polyfill embedded via the content itself.

- https://github.com/mdn/content/blob/main/files/en-us/web/mathml/element/munderover/index.md?plain=1#L29

This is what an example will look like without the above CSS polyfill and font:

![Shows a MathML example rendered in the browser. The text is rather small and bold.](https://user-images.githubusercontent.com/10350960/195450102-5530f215-6d7f-4b63-880e-9b1be887c846.png)

This is the same example with the second polyfill loaded:

![The same MathML example rendered with a different polyfill. The text is rendered larger and less bold.](https://user-images.githubusercontent.com/10350960/195450411-4dbbf491-741d-402f-a27f-c08a838f8fea.png)

The question then is, which of the two polyfills should we be using. Should we swap out the CSS and font face for the one loaded from @fred-wang or, remove the second polyfill and stick with the current polyfill in Yari. To be clear, if we decide to go with Fred's, we will update the component in Yari to use Fred's CSS and font as opposed to adding it in the content.

Let me know your thoughts. Thanks.

**Comment by fred-wang** [comment]
So to be clear, the inline stylesheet I inserted is not strictly speaking a polyfill, but just a stylesheet to load a Web font (Latin Modern Math) for MathML.

IIUC, the CSS polyfill is only loaded when MathML is not supported by the browser (and the page contains MathML) per https://github.com/mdn/yari/blob/80a3a9876b41d903d678636d88cb6d84a100c7b2/client/src/document/mathml-polyfill/index.tsx#L26

It is based on a CSS polyfill I wrote for browsers not supporting MathML, but has additional rule to load a Web font (STIX Two Math): https://github.com/mdn/yari/blob/80a3a9876b41d903d678636d88cb6d84a100c7b2/client/src/document/mathml-polyfill/mathml.css#L5

I believe we want to provide a Web font for MathML in any case (either Latin Modern Math or STIX Two Math) as a fallback when the system does not have Web fonts pre-installed. The CSS polyfill on the other hand is only needed for browsers without MathML support and shouldn't be loaded for browsers that do support MathML.

So I guess the right thing to do is to have:

1. Load a stylesheet with STIX Two Math as a Web fonts (probably WOFF2 instead of OTF) for pages containing any `<math>` tag.
2. Keep the existing MathML polyfill for browsers without MathML support (but remove the @font-face for STIX Two Math since that would be done in 1)
3. Remove any inline `<link>` to https://fred-wang.github.io/MathFonts/

**Reply by schalkneethling** [reply]
Thank you for your feedback and suggestions here. This is valuable.

**Reply by bsmth** [reply]
>So to be clear, the inline stylesheet I inserted is not strictly speaking a polyfill, but just a stylesheet to load a Web font (Latin Modern Math) for MathML.

I renamed the discussion to reflect this, in case anyone is wondering :)

**Comment by bsmth** [comment]
I think this makes sense. Currently, we have places like the [following section on `munderover`](https://github.com/mdn/content/blob/main/files/en-us/web/mathml/element/munderover/index.md?plain=1#L29-L34)

````markdown
```html hidden
 <link
   rel="stylesheet"
   href="https://fred-wang.github.io/MathFonts/LatinModern/mathfonts.css"
  />
```

```html
<math display="block">
  <munderover>
    <mo>∑</mo>
    <mrow>
      <mi>n</mi>
      <mo>=</mo>
      <mn>1</mn>
    </mrow>
    <mrow>
      <mo>+</mo>
      <mn>∞</mn>
    </mrow>
  </munderover>
</math>
```
````

After including the CSS fonts in Yari at build time, this would be simpler to write as:

````markdown
```html
<math display="block">
  <munderover>
    <mo>∑</mo>
    <mrow>
      <mi>n</mi>
      <mo>=</mo>
      <mn>1</mn>
    </mrow>
    <mrow>
      <mo>+</mo>
      <mn>∞</mn>
    </mrow>
  </munderover>
</math>
```
````

Note that it might be possible to detect pages that need this asset included [via the page tags](https://github.com/mdn/content/blob/main/files/en-us/web/mathml/element/munderover/index.md?plain=1#L5).

**Comment by Rumyra** [comment, marked as answer]
This sounds like a good plan to me:

> So I guess the right thing to do is to have:
> Load a stylesheet with STIX Two Math as a Web fonts (probably WOFF2 instead of OTF) for pages containing any <math> tag.
> Keep the existing MathML polyfill for browsers without MathML support (but remove the https://github.com/font-face for STIX Two Math since that would be done in 1)
> Remove any inline <link> to https://fred-wang.github.io/MathFonts/

I'm going to open an issue on yari for it

**Reply by Rumyra** [reply]
https://github.com/mdn/yari/issues/7607

**Reply by bsmth** [reply]
And https://github.com/mdn/yari/pull/7620

**Comment by bsmth** [comment]
There's been a change that's landed recently for this: https://github.com/mdn/yari/pull/7620

**Comment by fred-wang** [comment]
Analysis for the `files/en-us/web/mathml/` (there are more usages to check):

- files/en-us/web/mathml/element/math/index.md: A math font is at least needed for the big summation symbol.
- files/en-us/web/mathml/element/mfrac/index.md: A math font is at least needed for the stretched fences.
- files/en-us/web/mathml/element/mover/index.md: A math font is at least needed for the stretched bracket.
- files/en-us/web/mathml/element/mroot/index.md: not sure why a math font is needed, maybe to provide a glyph for the radical sign or the italic x.
- files/en-us/web/mathml/element/msqrt/index.md: not sure why a math font is needed, maybe to provide a glyph for the radical sign or the italic x.
- files/en-us/web/mathml/element/mstyle/index.md: A math font is at least needed for the big operators (sum & product).
- files/en-us/web/mathml/element/msup/index.md: not sure why a math font is needed, maybe to provide a glyph for italic X.
- files/en-us/web/mathml/element/munder/index.md: A math font is at least needed for the stretched bracket.
- files/en-us/web/mathml/element/munderover/index.md: A math font is at least needed for the stretched bracket.

It seems none of them needs Latin Modern Math specifically and the page already uses `https://fred-wang.github.io/MathFonts/STIX/mathfonts.css` (although I'm not sure where that's coming from). So we can remove these. 

The ones in `files/en-us/web/mathml/element/munderover/authoring/index.md` are used as code samples and should be kept.

**Reply by fred-wang** [reply]
> It seems none of them needs Latin Modern Math specifically and the page already uses https://fred-wang.github.io/MathFonts/STIX/mathfonts.css (although I'm not sure where that's coming from). 

Just replying to myself here: the stylesheet for STIX Two Math is loaded by `kumascript/src/live-sample.ts` after https://github.com/mdn/yari/pull/7621
